### PR TITLE
Feature/teams pages

### PR DESCRIPTION
**Pull request by vince-a-m**
- Added UserList component for displaying users
- Added AddUserDialog for user selection
- Added TeamsList and TeamsView components for displaying all teams and an individual team
- Added a CreateDialog component for getting inputs for a new object from the user.
- Added placehold REST API routes for CRUD operations on teams

### CODE REVIEW

**Review by clemiller — CHANGES_REQUESTED** [review]
This is a great start to the Teams interface! The overall workflow of creating/editing teams and updating the user lists looks good. I've made a few comments below on some suggested improvements. A couple of other thoughts:

- [ ] I haven't looked through the code smells reported by sonarcloud, so if you haven't already, it may be worth looking into to see if some of those could be resolved.
- [ ] Function docs - would you mind adding these to explain what each function does? It'll make it easier for others to understand each function's purpose and how they're used
- [ ] The spacing around the teams header and the "create a new team" button is a little tight in comparison to the other Workbench pages. Similar spacing issues with the list of users on the team view page:
image
image

File: app/src/app/classes/authn/team.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
+import { Observable, of } from 'rxjs';
+import { map, switchMap } from 'rxjs/operators';
+import { RestApiConnectorService } from '../../services/connectors/rest-api/rest-api-connector.service';
+import { logger } from '../../util/logger';
+import { Serializable, ValidationData } from '../serializable';
+import { UserAccount } from './user-account';
```
[the remark this answers is not in the document]
Comment (vince-a-m) [reply app/src/app/classes/authn/team.ts:6]: They are going to be needed once the teams actually exist, if they aren't used after that I'll remove them.

File: app/src/app/classes/authn/team.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+        /*if ('created' in raw) {
+            if (typeof (raw.created) === "string") this.created = new Date(raw.created);
+            else logger.error("TypeError: created field is not a string:", raw.created, "(", typeof (raw.created), ")")
+        } else { this.created = new Date(); }
+
+        if ('modified' in raw) {
+            if (typeof (raw.modified) === "string") this.modified = new Date(raw.modified);
+            else logger.error("TypeError: modified field is not a string:", raw.modified, "(", typeof (raw.modified), ")")
+        } else { this.modified = new Date(); }*/
```
[the remark this answers is not in the document]
Comment (vince-a-m) [reply app/src/app/classes/authn/team.ts:74]: Yes

File: app/src/app/views/admin-page/admin-page.component.html [hunk 3]
```diff
@@ -26,15 +26,30 @@ <h3 class="display-icon">
...
-                    </h2>
-                    <h3 class="display-icon">
-                        <mat-icon>people_alt</mat-icon>
-                    </h3>
+                <button mat-raised-button color="primary" routerLink="teams">
+                  <h2>
+                    <br>Teams
+                  </h2>
+                  <h3 class="display-icon">
+                      <mat-icon>groups</mat-icon>
```
[the remark this answers is not in the document]
Comment (vince-a-m) [reply app/src/app/views/admin-page/admin-page.component.html:34]: :)

File: app/src/app/classes/authn/team.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    /**
+     * Validate the current object state and return information on the result of the validation
+     * @param {RestApiConnectorService} restAPIService: the REST API connector through which asynchronous validation can be completed
+     * @returns {Observable<ValidationData>} the validation warnings and errors once validation is complete.
+     */
+    public validate(restAPIService: RestApiConnectorService): Observable<ValidationData> {
+        const validation = new ValidationData();
+        const obs = new Observable<ValidationData>();
+        return obs;
+    }
```
[the remark this answers is not in the document]
Comment (vince-a-m) [reply app/src/app/classes/authn/team.ts:86]: Will add once there API endpoints exist

File: app/src/app/classes/authn/team.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    /**
+     * Save the current state of the STIX object in the database. Update the current object from the response
+     * @param restAPIService [RestApiConnectorService] the service to perform the POST/PUT through
+     * @returns {Observable} of the post
+     */
+     public save(restAPIService: RestApiConnectorService): Team {
+        const putObservable = restAPIService.putTeam(this);
+        return putObservable;
+    }
```
[the remark this answers is not in the document]
Comment (vince-a-m) [reply app/src/app/classes/authn/team.ts:96]: Yeah, I'm just mocking the API for the time being so I did what I had to do to get it run.  Will change once the API exists

File: app/src/app/components/users-list/users-list.component.html
[the remark this answers is not in the document]
Comment (vince-a-m) [reply app/src/app/components/users-list/users-list.component.html:1]: @clemiller Sorry, I don't know what this means.  Could you elaborate?

**Comment by vince-a-m** [comment]
A lot of the code smells are due to weird hacky workarounds related to the API not actually existing yet.  

Can do with adding the comments to the functions 👍

File: app/src/app/views/admin-page/teams/teams-view-page/teams-view-page.component.html [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  <div class="row">
+    <div class="col">
+        <!-- DESCRIPTION -->
+        <app-descriptive-property [config]="{
+            mode: editing? 'edit' : 'view',
+            object: team,
+            field: 'description',
+            firstParagraphOnly: false, 
+            label: 'Description'
+        }"></app-descriptive-property>
```
[the remark this answers is not in the document]
Comment (vince-a-m) [reply app/src/app/views/admin-page/teams/teams-view-page/teams-view-page.component.html]: Made a fix, might need some work

**Review by clemiller — CHANGES_REQUESTED** [review]
I ran into a couple of errors while I was testing out the teams interface:

- An error is thrown when a team is created without a description, as well as after clicking "edit" on a team without a description:

image

- Errors are thrown and the API crashes when an existing team is edited and given a non-unique name. The team name should be validated before sending the PUT request to the API and prevent the user from saving if it's not unique.

Also, if there are a lot of users in a Workbench instance, it might take a while to find all of the users to add to a team. Would it be possible to add searching for users in the add-dialog window?
image

**Pull request merged**